### PR TITLE
Proxy Agent Shared: Introduce custom result and error types

### PR DESCRIPTION
**Pull request by mayankdaruka-msft**
This PR is addressing the feedback from #117.

Similar to this PR - [Proxy Agent: Introduce custom result and error types](https://github.com/Azure/GuestProxyAgent/pull/134) - but for the shared crate. No subsequent PRs will be required here.

### CODE REVIEW

File: proxy_agent_shared/src/error.rs [hunk 1]
```diff
@@ -0,0 +1,64 @@
+#[derive(Debug, thiserror::Error)]
```
Comment (ZhidongPeng) [inline proxy_agent_shared/src/error.rs:4]: // Copyright

File: proxy_agent_shared/src/error.rs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    ParseVersion(ParseVersionErrorType),
+
+    #[error("Findmnt command {0}")]
+    Findmnt(String),
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum ParseVersionErrorType {
+    #[error("Invalid version string")]
+    InvalidString,
```
Comment (ZhidongPeng) [inline proxy_agent_shared/src/error.rs]: we properly need add the input `version-string` to the error message.

File: proxy_agent_shared/src/result.rs [hunk 3]
```diff
@@ -0,0 +1,3 @@
+use super::error::Error;
```
Comment (ZhidongPeng) [inline proxy_agent_shared/src/result.rs:4]: // Copyright

Comment (ZhidongPeng) [reply proxy_agent_shared/src/result.rs:4]: Please help add `// Copyright` to proxy_agent/commom/result.rs file too

File: proxy_agent_shared/src/error.rs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    Json(#[from] serde_json::Error),
+
+    #[error("Failed to create regex with error: {0}")]
+    Regex(#[from] regex::Error),
+
+    #[error("{0}")]
+    ParseVersion(ParseVersionErrorType),
+
+    #[error("Findmnt command {0}")]
+    Findmnt(String),
```
Comment (ZhidongPeng) [inline proxy_agent_shared/src/error.rs]: rename it to `CommandError` like?
` #[error("{0} command: {1}")]
    CommandError(String, String),`

**Review by ZhidongPeng — CHANGES_REQUESTED** [review]

File: proxy_agent_shared/src/error.rs (hunk 2 above)
Comment (mayankdaruka-msft) [reply proxy_agent_shared/src/error.rs]: Added, thanks.

**Review by ZhidongPeng — APPROVED** [review]

**Pull request merged**